scripts/src/precheck/submission.py

```diff
@@ -0,0 +1,220 @@
+import os
+import requests
+import semver
+
+from dataclasses import dataclass, field
+
+from checkprcontent import checkpr
+
+xRateLimit = "X-RateLimit-Limit"
+xRateRemain = "X-RateLimit-Remaining"
+
+
+class SubmissionError(Exception):
+    """Root Exception for handling any error with with the submission"""
+
+    pass
+
+
+class DuplicateChartError(SubmissionError):
+    """This Exception is to be raised when the user attempts to submit a PR with more than one chart"""
+
+    pass
+
+
+class VersionError(SubmissionError):
+    """This Exception is to be raised when the version of the chart is not semver compatible"""
+
+    pass
+
+
+class OwnersFileError(SubmissionError):
+    """This Exception is to be raised when the user incorrectly attempts to submit an OWNER file"""
+
+    pass
+
+
+@dataclass
+class Chart:
+    category: str = None
+    organization: str = None
+    name: str = None
+    version: str = None
+
+    def register_chart_info(self, category, organization, name, version):
+        if (
+            (self.category and self.category != category)
+            or (self.organization and self.organization != organization)
+            or (self.name and self.name != name)
+            or (self.version and self.version != version)
+        ):
+            msg = "[ERROR] A PR must contain only one chart. Current PR includes files for multiple charts."
+            print(msg)
+            raise DuplicateChartError(msg)
+
+        if not semver.VersionInfo.isvalid(version):
+            msg = (
+                f"[ERROR] Helm chart version is not a valid semantic version: {version}"
+            )
+            print(msg)
+            raise VersionError(msg)
+
+        self.category = category
+        self.organization = organization
+        self.name = name
+        self.version = version
+
+
+@dataclass
+class Report:
+    found: bool = False
+    signed: bool = False
+    path: str = None
+
+
+@dataclass
+class Source:
+    found: bool = False
+    path: str = None  # Path to the Chart.yaml
+
+
+@dataclass
+class Tarball:
+    found: bool = False
+    path: str = None
+    provenance: str = None
+
+
+@dataclass
```

Comment:
Because this is a dataclass, I would understand that all of the class variables defined would be settable by default (in the generated init function). When I'm running this locally, it feels like I generally only want to pass in api_url, and the rest of the fields _should_ be derived.

Do I have a bad read on the application of this class, or should the rest of these class variables not be settable?

Reply:
Good question.

TLDR: A constructor that accepts *all* attributes is needed for the JSONDecoder to function. Specifically [here](https://github.com/openshift-helm-charts/development/pull/332/files/5a55ce326e4cde2a43d0f5ad634d7604afaea6c9#diff-ded7b02fcbdf810b9bf20e8f306d88f09f94202e4b3cf8a49149b55ab05bfcb4R39).

I envision two cases where we are going to instantiate a `Submission` object:
1. In the "pre-check" job (we can revisit the name of this job BTW), we are given an api_url, and the rest is derived. At the end of this process, we dump the Submission object as JSON to an [artifact file](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts#passing-data-between-jobs-in-a-workflow).
2. In following jobs, we want to access some of the Submission information, such as the path of the report, or the category. Either we do that directly in the jobs shell (`run`) using `jq`, or we have a python script that loads the artifact file. The later can be easily achieved with the `JSONDecoder` (Note that this is needed due to the nested objects: `Chart`, `Report`, `Source` and `Tarball`).

In order to know if we're dealing with 1. or 2., we check in `__post_init__` if we were provided with a list of `modified_files`. This way we don't populate the class again from scratch and override the given input data (e.g. artifact file) - though of course it should find the exact same info.

That's what I was trying to explain in [this comment](https://github.com/openshift-helm-charts/development/pull/332/files/5a55ce326e4cde2a43d0f5ad634d7604afaea6c9#diff-6cdd6c8509751e3be6c04f9c38d4c43c16760aab0318c9ed6cbad88bc9b89300R100-R102) that you didn't understand. I don't blame you, I haven't done an amazing job at explaining :) I will rework the comment.